File: 496-next-greater-element-1/Solution.java
// 这题关键是要清楚第二个数组各个元素next great element的结构
public class Solution {
    public int[] nextGreaterElement(int[] findNums, int[] nums) {
        Map<Integer, Integer> map = new HashMap<>();
        Stack<Integer> stack = new Stack<>();
        for (int num : nums) {
            while (!stack.isEmpty() && stack.peek() < num)
                map.put(stack.pop(), num); // 这里只能用pop，不能用peek
            stack.push(num);
        }
        for (int i = 0; i < findNums.length; i++)
            findNums[i] = map.getOrDefault(findNums[i], -1);
        return findNums;
    }
}

File: 496-next-greater-element-1/solution.py
class Solution(object):
    def nextGreaterElement(self, findNums, nums):
        """
        :type findNums: List[int]
        :type nums: List[int]
        :rtype: List[int]
        """
        map, stack = {}, []  # using list as a stack
        # stack里面存的都是暂时没有找到next greater element的元素
        for num in nums:
            while stack and stack[-1] < num:
                map[stack.pop()] = num
            stack.append(num)
        res = []
        for num in findNums:
            res.append(map.get(num, -1))
        return res